### PR TITLE
EZP-24213: FullText stopWordThreshold should be percentage of content count

### PR DESCRIPTION
**Pull request by andrerom**
TL;DR; FullText was hardcoded to disable search on phrases that exists in more then 20 content objects..
See issue for details: https://jira.ez.no/browse/EZP-24213

*Approach: In original discussion with @pspanja I suggested exposing a spiContentHandler->getTotalCount() method to provide total count and cache it in SPI cache. I have initial work on that, however ended up just caching it locally on FullTextHandler as that turned out to much simpler and a count query is compared to everything else we do here very fast. If someone see some immediate value also other places in having spiContentHandler->getTotalCount() I can re add it, or do it as followup.*

### CODE REVIEW

**Comment by pspanja** [comment]
This looks good to me, but maybe we should mention it in BC notes, as the config is exposed. We could also keep both `searchThresholdValue` and `searchThresholdFactor`, handling it in the same way Legacy Stack does.

**Comment by andrerom** [comment]
> We could also keep both searchThresholdValue and searchThresholdFactor, handling it in the same way Legacy Stack does.

From my understanding in legacy, StopWordThresholdValue has opposite meaning, so will disable stop words completely if number of objects are below this. I skipped this for new stack as I felt the majority of cases even with 100 objects you have a benefit from this if you search for something used by all of them. So others will have to configure instead. Sounds reasonable?

But BC doc can be added indeed, I didn't know it was possible to inject it given the other PR: #1166

**Comment by andrerom** [comment]
review ping besides bc doc topic: @bdunogier @lolautruche @pspanja @yannickroger

**Comment by bdunogier** [comment]
It looks good to me, but I gotta admit I don't feel very comfortable there :-)

**Comment by andrerom** [comment]
What specifically is unclear? Anything we can describe to make it more clear? :)

**Comment by andrerom** [comment]
Added more php doc now to hopefully make this feature a bit more clear. The feature is specific to Legacy Search Engine, we had similar feature in legacy (see site.ini).

**Comment by bdunogier** [comment]
I basically don't know squat about the legacy search engine, within a legacy context at least. I'll read again, slower this time :-)

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php [hunk 1]
```diff
@@ -223,5 +240,43 @@ public function handle( CriteriaConverter $converter, SelectQuery $query, Criter
...
+        $query
+            ->select(
+                $query->alias( $query->expr->count( '*' ), 'count' )
+            )
+            ->from( $this->dbHandler->quoteTable( "ezcontentobject" ) );
+
+        $statement = $query->prepare();
+        $statement->execute();
+
+        // Makes sure factor is between 0 and 1
```
Comment (bdunogier) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: Shouldn't we ensure that when we inject the config value into the object ?

**Comment by bdunogier** [comment]
Yep, makes much more sense. Good one.

About the config, @andrerom, i'm almost sure it was not, but @pspanja should confirm that.

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php (hunk 1 above)
Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: yes ideally, I kind of just added this for convenience, but can change to throw in ctor

Comment (bdunogier) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: ctor sounds nice. It should end up in the semantic config validation as
well, ideally (but it's a bonus).

Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: fixed

**Comment by pspanja** [comment]
@andrerom, @bdunogier 
> Was this possible to inject via yml config? (if so I'll create bc note)

Not in semantic config, but it was possible to redifine the service adding your own configuration, which would be merged with default one. See https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php#L90. That is what I meant in https://github.com/ezsystems/ezpublish-kernel/pull/1166.

**Comment by pspanja** [comment]
> From my understanding in legacy, StopWordThresholdValue has opposite meaning, so will disable stop words completely if number of objects are below this. I skipped this for new stack as I felt the majority of cases even with 100 objects you have a benefit from this if you search for something used by all of them. So others will have to configure instead. Sounds reasonable?

That's my understanding as well.. this is ok with me then :+1:

**Comment by andrerom** [comment]
> Not in semantic config, but it was possible to redifine the service adding your own configuration, which would be merged with default one. See https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php#L90. That is what I meant in #1166.

ok, bc doc it is then :)

**Comment by andrerom** [comment]
bc doc added, ready for review!

File: doc/bc/changes-5.3.md [hunk 2]
```diff
@@ -102,6 +102,17 @@ Changes affecting version compatibility with former or future versions.
...
+* 5.3.5: Legacy Search Engine FullText searchThresholdValue -> stopWordThresholdFactor
+
+    EZP-24213: the "Stop Word Threshold" configuration, `searchThresholdValue`, was hardcoded
+    to 20 items. It is now changed to `stopWordThresholdFactor`, a factor (between 0 and 1)
+    for the percentage of content objects to set the Stop Word Threshold to. Default value
+    is set to 0.66, meaning if you search for a common word like "the", it will be ignored
+    from the search expression if more then 66% of your content contains the word.
+
+    Note: Does not affect future Solr/ElasticSearch search engines which has far more
+          advance search options built in.
```
Comment (bdunogier) [inline doc/bc/changes-5.3.md]: advanced (+d)

File: doc/bc/changes-6.0.md [hunk 3]
```diff
@@ -54,6 +54,17 @@ Changes affecting version compatibility with former or future versions.
...
+* Legacy Search Engine FullText searchThresholdValue -> stopWordThresholdFactor
+
+    EZP-24213: the "Stop Word Threshold" configuration, `searchThresholdValue`, was hardcoded
+    to 20 items. It is now changed to `stopWordThresholdFactor`, a factor (between 0 and 1)
+    for the percentage of content objects to set the Stop Word Threshold to. Default value
+    is set to 0.66, meaning if you search for a common word like "the", it will be ignored
+    from the search expression if more then 66% of your content contains the word.
+
+    Note: Does not affect future Solr/ElasticSearch search engines which has far more
+          advance search options built in.
```
Comment (bdunogier) [inline doc/bc/changes-6.0.md]: same as above

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php [hunk 4]
```diff
@@ -178,18 +201,26 @@ protected function getWordIdSubquery( SelectQuery $query, $string )
             $wordExpressions[] = $this->getWordExpression( $subQuery, $token );
         }
 
+        // Create where expression
```
Comment (bdunogier) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: I'd remove that, the variable name is clear enough.

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php [hunk 5]
```diff
@@ -178,18 +201,26 @@ protected function getWordIdSubquery( SelectQuery $query, $string )
...
+        $whereCondition = $subQuery->expr->lOr( $wordExpressions );
+
+        // If stop word threshold is below 100% we make sure it is applied
+        // @see getStopWordThresholdValue()
+        if ( $this->configuration['stopWordThresholdFactor'] < 1 )
+        {
+            $whereCondition = $subQuery->expr->lAnd(
```
Comment (bdunogier) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: hmmm, is this right that `$whereCondition` is overwritten right after it has been defined ?

Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: see one line below

Comment (bdunogier) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: Hmmmm, OK. Didn't see that at first glance, sorry.

File: doc/bc/changes-5.3.md (hunk 2 above)
Comment (andrerom) [inline doc/bc/changes-5.3.md]: fixed

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php (hunk 4 above)
Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: fixed

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php (hunk 5 above)
Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FullText.php]: I updated the inline comment to maybe make it more clear

**Comment by pspanja** [comment]
Looks good to me now, +1.

**Comment by bdunogier** [comment]
+1

**Comment by andrerom** [comment]
Thanks for the reviews, :ship:ing ;)
